source/server/debug_session_properties_restricted_service.cpp

```diff
@@ -0,0 +1,244 @@
+  IsDebugSessionServerOutOfProcessResponse* response)
+{
+  bool out_of_process = false;
+  auto status = get_bool_property((NISysCfgResourceProperty)0x10003000, request->product_id(), request->serial_number().c_str(), &out_of_process);
+  response->set_out_of_process(out_of_process);
+  return status;
+}
+
+::grpc::Status DebugSessionPropertiesRestrictedService::SetDebugSessionEnabled(
+  ::grpc::ServerContext* context,
+  const SetDebugSessionEnabledRequest* request,
+  SetDebugSessionEnabledResponse* response)
+{
+  uint32_t enabled = request->enabled() ? 1 : 0;
+  auto status = set_uint_property((NISysCfgResourceProperty)0x10002000, request->product_id(), request->serial_number().c_str(), enabled);
+  return status;
+}
+
+::grpc::Status DebugSessionPropertiesRestrictedService::SetDebugSessionServerOutOfProcess(
+  ::grpc::ServerContext* context,
+  const SetDebugSessionServerOutOfProcessRequest* request,
+  SetDebugSessionServerOutOfProcessResponse* response)
+{
+  auto status = set_bool_property((NISysCfgResourceProperty)0x10003000, request->product_id(), request->serial_number().c_str(), request->out_of_process());
+  return status;
+}
+
+::grpc::Status DebugSessionPropertiesRestrictedService::get_bool_property(
+  NISysCfgResourceProperty property_id,
+  uint32_t product_id,
+  const char* serial_number,
+  bool* value)
+{
+  NISysCfgStatus status = NISysCfg_OK;
+  NISysCfgSessionHandle session = NULL;
+  NISysCfgFilterHandle filter = NULL;
+  NISysCfgEnumResourceHandle resources_handle = NULL;
+  NISysCfgResourceHandle resource = NULL;
+  NISysCfgBool syscfg_value = NISysCfgBoolFalse;
+
+  try {
+    auto library = get_syscfg_library_interface();
+    if (NISysCfg_Succeeded(status = open_or_get_localhost_syscfg_session(&session))) {
+      if (NISysCfg_Succeeded(status = library->CreateFilter(session, &filter))) {
+        library->SetFilterProperty(filter, NISysCfgFilterPropertyIsDevice, NISysCfgBoolTrue);
+        library->SetFilterProperty(filter, NISysCfgFilterPropertyProductId, product_id);
+        library->SetFilterProperty(filter, NISysCfgFilterPropertySerialNumber, serial_number);
+        if (NISysCfg_Succeeded(status = library->FindHardware(session, NISysCfgFilterModeMatchValuesAll, filter, NULL, &resources_handle))) {
+          if (NISysCfg_Succeeded(status) && (status = library->NextResource(session, resources_handle, &resource)) == NISysCfg_OK) {
+            library->GetResourceProperty(resource, property_id, &syscfg_value);
+            *value = syscfg_value != NISysCfgBoolFalse;
+            library->CloseHandle(resource);
+          }
+          library->CloseHandle(resources_handle);
+        }
+        library->CloseHandle(filter);
+      }
+    }
+  }
+  catch (nidevice_grpc::NonDriverException& ex) {
+    return ex.GetStatus();
+  }
+
+  if (NISysCfg_Failed(status)) {
+    return ::grpc::Status(::grpc::StatusCode::INTERNAL, kDebugSessionPropertyAccessFailedMessage);
```

Comment:
This is not how we do this now. It should follow the convention of:
- Returning UNKNOWN status and the real already-looked-up error message in the Status object.
- Putting the actual status code in the trailer

Reply:
After talking with you, I asked Harsha about this.  He said that returning a static error message with the INTERNAL status for SysCfg operations was an intentional decision since the user is not attempting to interact with SysCfg but just perform the operation.  Providing detailed SysCfg errors in this context is not useful.  While this is a restricted proto file, I think these methods fulfill a similar function and should follow the same convention.

Reply:
After further discussion, I've updated as suggested.